Add resource tweak (#39)
sanger-pathogens/nextflow-commons: configs/README.md

# Configuring Our Pipeline with Custom Resources

This README explains how to configure PaM Nextflow pipelines with custom resources (`CPU`, `memory`, `queue`, and `runtime`).
Examples use the `generate_mags` pipeline, but the same principles apply to all pipelines.

---

## 1. Download the `resources_tweak.config` template

The template resource_tweak.config provides examples for adjusting process-specific resources in PaM pipelines. You can use it as a starting point for your own resource configuration. To download use the below command in your working directory.

```bash
wget https://raw.githubusercontent.com/sanger-pathogens/nextflow-commons/refs/heads/master/configs/resource_tweak.config 
```

## 2. Identify processes for each pipeline

> Each Nextflow pipeline is built from processes (tasks). To customise resources, you first need to know which processes are used in your pipeline. There are two ways to find them:

Reviewer: ```suggestion
Each Nextflow pipeline is built from processes, which are applied to each input data item (task). To customise resources, you first need to know which processes are used in your pipeline. There are two ways to find them:
```

### Using our gitlab repositories: 

> 1) Go to the relevant GitLab repository for your pipeline. ([pathogen pipelines gitlab page](https://gitlab.internal.sanger.ac.uk/sanger-pathogens/pipelines))

Reviewer: ```suggestion
1) Go to the relevant git repository for your pipeline. ([pathogen pipelines gitlab page](https://gitlab.internal.sanger.ac.uk/sanger-pathogens/pipelines))
```

Each process in a Nextflow pipeline has default resource settings. These control how many CPUs, how much memory, which queue, and how long the job can run. You may see these resources defined in different ways depending on the pipeline.

In all examples we will submit `ASSEMBLY` to the `normal` queue requesting `32GB` of memory and having a maximum time of `8 hours` to run.

1. Directly inside the process:

Some processes specify their resources explicitly in the process block:
```bash

process ASSEMBLY {
    cpus = 8
    memory = 32GB
    queue = normal 
    time = 8.h
}
```
2. Using labels from nextflow-commons:

Many pipelines use labels instead of hard-coding resources within the process block. These labels map to definitions in the [nextflow-commons.config](https://github.com/sanger-pathogens/nextflow-commons/blob/master/configs/nextflow.config) (see *section 8 "Understanding nextflow-commons labels"* for examples of label resolution)
```bash

process ASSEMBLY {
    label 'cpu_8'
    label 'mem_32'
    label 'time_queue_from_normal'
}
```

3) Using nextflow.config (in the main pipeline repository)

```bash
process { 

>     withName ASSEMBLY {

Reviewer: ```suggestion
    withName: ASSEMBLY {
```

            cpus = 8
            memory = 32GB
            queue = normal 
            time = 8.h
    }
}
```

**Tip:** When checking defaults:
1. Look in the process definition (`.nf` file)

> 2. Check if labels are used, and trace them back to the [nextflow-commons.config](https://github.com/sanger-pathogens/nextflow-commons/blob/master/configs/nextflow.config)

Reviewer: ```suggestion
2. Check if labels are used, and trace them back to the [nextflow-commons.config](./nextflow.config)
```

3. Check `nextflow.config` in the main pipeline repository

## 4. Priority:

Nextflow applies resource settings according to priority, with the highest-priority value taking effect if the same parameter is defined in multiple places. This allows you to override resource parameters. 

### Priorities low -> high:

1) Process definition (`.nf` file)
2) nextflow.config in the pipeline repository
3) Custom config passed with `-c` (e.g. resource_tweak.config) – highest priority and applied last.

## 5. How to override the resource requirements for your process of interest:

1) Edit the resource_tweak.config file by adding a process block with the exact name of the process as defined in the .nf file 
2) Specify the resource parameters you want to override.

Example:

```bash
process { 
    withName PROCESS_NAME {
        cpus = 8
        memory = 32GB
        queue = normal 
        time = 8 hr
    }

>     withName PROCESS_NAME2 {

Reviewer: ```suggestion
    withName: PROCESS_NAME2 {
```

### Memory and queue conflicts:

Jobs cannot request memory < 196 GB and submit to the hugemem queue.
Jobs cannot request memory < 745 GB and submit to the teramem queue.

### Label and parameter conflicts with examples:

If a parameter is not specified in `resource_tweak.config` for a process, the value will be inherited from `nextflow.config` as the next default, and then from the process-level defaults if not specified there. 

### Process script conflicts: 

Some scripts within processes may include resource arguments that conflict with your specified parameters. For example, in the `Jellyfish Generator` pipeline, the process `JELLY_GEN` has `memory` hard-coded to `50 GB` and `CPUs` set to `8`:

```bash
jellyfish count /dev/fd/0 -m 50 -s 100M -o ${db_name} -t 8 -C
```
To override this behavior, you would need to clone the pipeline and manually edit the command inside the process script.

## 6. Examples of overriding resources: 

The below section will show examples of how to overwrite the resource parameters and retry straties. 

All examples bellow will follow the ASSEMBLY process from generate mags.

> **Process:**

Reviewer: ```suggestion
**Process, as defined in the `*.nf` module file:**
```

```bash

process ASSEMBLY {
    label 'cpu_8'
    label 'mem_32'
    label 'time_queue_from_normal'
}
```

> **Example 1:** Override all parameters

Reviewer: ```suggestion
**Example 1:** Override all resource parameters
```

> Using resource_tweak.config to override all resources at the process level:

Reviewer: ```suggestion
Using `resource_tweak.config` to override all resources at the process level:
```

```bash
process {
    withName: 'ASSEMBLY' {
        cpus = 4
        memory = 50.GB
        queue = 'long'
        time = 48.h
    }
}
```
**Result:** `ASSEMBLY` will run with `4 CPUs`, `50 GB memory`, on the `long queue`, for up to `48 hours`.`

**Example 2:** Override only some parameters

Using resource_tweak.config to some resources at the process level:
```bash
process {
    withName: 'ASSEMBLY' {
        memory = 50.GB
        queue = 'long'
        time = 48.h
    }
}
```
**Result:** CPUs remain from the process label (`cpu_8` → `8 CPUs`).`memory`, `queue`, and `time` are overridden.

**Example 3:** Partial override with label fallback

```bash
process {
    withName: 'ASSEMBLY' {
        memory = 50.GB
        queue = 'long'
    }
}
```

> **Result:** Even though `queue = long` has been supplied, because `time ` parameter has not been specified, the default label `'time_queue_from_normal'` from the `nextflow-commons.config`will be applied. This label overrides the queue setting, and the job will be submitted to the normal queue. Refer to *Section 8 "Understanding nextflow-commons labels"*, for details.

Reviewer: ```suggestion
**Result:** Maximum task job runtime remains 12h. Why? Even though `queue = long` has been specified in the `resource_tweak.config` file, the `time ` parameter has not been modified there, and takes its value from the default process label `'time_queue_from_normal'`, which is interpreted as per the `nextflow-commons.config`. Applying this label overrides the queue setting, and the job will be submitted to the normal queue, and accordingly sets the task paramter `time` value to 12 . Refer to *Section 8 "Understanding nextflow-commons labels"*, for more details.
```

> **Example 4:** Memory resource escalation

Reviewer: ```suggestion
### **Example 4:** Changing memory resource escalation strategy
```

```bash
process {
    withName: 'ASSEMBLY' {
        cpus    = 8
        memory  = { check_max( escalate_exp( 64.GB, task, 2 ), 'memory' ) }
        queue = 'long'
        time    = 48.h
    }
}
```
Ensure to add the folowing functions copied from the `nextflow-commons.config` to the `resource_tweak config`: 
1) `def check_max`
2) `def escalate_exp`

**Result:** The memory resource will start at `64 GB` on the first attempt.

> On each retry (default 2 retries), the memory will double (`64` → `128` → `256 GB`, etc.) depending on the multiplier (in this case 2) .
> Check_max ensures that the escalated memory does not exceed farm/system limits.

Reviewer: ```suggestion
On each retry (default 2 retries), the memory will then double (64 → 128 → 256 GB, etc.) depending on the multiplier (in this case 2) .
The `check_max` function ensures that the escalated memory does not exceed farm/system limits.
```

## 7. Recommended resource combinations:

Resource maximums by default are: 

* `max_memory = 2.9.TB`
* `max_cpus = 256`
* `max_time = 720.h`

Always request resources sensibly according to [ISG farm documentation](https://ssg-confluence.internal.sanger.ac.uk/spaces/FARM/pages/101360547/Farm+documentation), and if in doubt please raise a ticket with PaM Info or ISG.

For detailed memory limits and runtime restrictions by queue, use:
```bash
bqueues -l queue_name
```

### Recommended resource combinations:

Note: The memory recommendations for each job are based on current ISG documentation but have not received official ISG approval.

| Queue       | Memory               | Runtime Limit |
|-------------|----------------------|---------------|
| **small**   | < 256 GB             | ≤ 30 min      |
| **normal**  | < 256 GB             | ≤ 12 hours    |
| **long**    | < 256 GB             | ≤ 48 hours    |
| **week**    | < 256 GB             | ≤ 168 hours   |
| **hugemem** | 256–745 GB           | ≤ 360 hours   |
| **teramem** | > 745 GB             | ≤ 360 hours   |
| **basement**| < 256 GB             | ≤ 720 hours   |

## 8. Understanding nextflow-commons labels: 

> Labels in Nextflow provide a central mechanism to define and manage resources for processes. To check what a label resolves to, refer to the [nextflow-commons.config](https://github.com/sanger-pathogens/nextflow-commons/blob/master/configs/nextflow.config). 

Reviewer: ```suggestion
Labels in Nextflow provide a central mechanism to define and manage resources for processes. To check what a label resolves to, refer to the [nextflow-commons.config](./nextflow.config). 
```

### Examples of label resolution

> label `'cpu_8'`: 

Reviewer: ```suggestion
#### label `'cpu_8'`: 
```

```bash
    withLabel:cpu_8 {
        cpus   = { check_max( 8, 'cpus' ) }
    }
```
This resolves to 8 cpus.

> label `'mem_32'`:

Reviewer: ```suggestion
#### label `'mem_32'`:
```

```bash
    withLabel:mem_32 {
        memory = { check_max( escalate_exp( 32.GB, task, 2 ), 'memory' ) }
    }
```
Resolves to `32 GB memory`.

> Enables retries, escalating exponentially: `32 GB` → `64 GB` → `128 GB`.

Reviewer: ```suggestion
- Enables retries, escalating exponentially: `32 GB` → `64 GB` → `128 GB`.
```

label `'time_30m'`:

```bash
    withLabel:time_30m {
        time   = { check_max( escalate_linear( 30.m, task ), 'time' ) }
    }
```

> Resolves to `30 minutes`.
> Enables retries with linear escalation: `30 m` → `60 m` → `90 m`, etc.

Reviewer: ```suggestion
- Resolves to `30 minutes`.
- Enables retries with linear escalation: `30 m` → `60 m` → `90 m`, etc.
```

> label `'time_queue_from_normal'`:

Reviewer: ```suggestion
#### label `'time_queue_from_normal'`:
```

## 6. Examples of overriding resources: 

The below section will show examples of how to override the resource parameters and retry strategies. 

All examples bellow will follow the ASSEMBLY process from generate mags.

**Process, as defined in the `*.nf` module file:**
```bash

process ASSEMBLY {
    label 'cpu_8'
    label 'mem_32'
    label 'time_queue_from_normal'
    
    [...]
}
```
### **Example 1:** Override all resource parameters

Using `resource_tweak.config` to override all resources for a specific process:
```bash
process {
    withName: 'ASSEMBLY' {
        cpus = 4
        memory = 50.GB
        queue = 'long'
        time = 48.h
    }
}
```
**Result:** `ASSEMBLY` will run with `4 CPUs`, `50 GB memory`, on the `long queue`, for up to `48 hours`.`

### **Example 2:** Override only some parameters

Using `resource_tweak.config` to override some resources for a specific process:
```bash
process {
    withName: 'ASSEMBLY' {
        memory = 50.GB
        queue = 'long'
        time = 48.h
    }
}
```
**Result:** `cpu` task parameter value remain as set from the process label (`cpu_8` → 8 CPUs).`memory`, `queue`, and `time` task parameter values are overridden.

> **Example 3:** Partial override with label fallback

Reviewer: ```suggestion
### **Example 3:** Partial override with label fallback
```

```bash
process {
    withName: 'ASSEMBLY' {
        memory = 50.GB
        queue = 'long'
    }
}
```
**Result:** Even though `queue = long` has been supplied, because `time ` parameter has not been specified, the default label `'time_queue_from_normal'` from the `nextflow-commons.config`will be applied. This label overrides the queue setting, and the job will be submitted to the normal queue. Refer to *Section 8 "Understanding nextflow-commons labels"*, for details.

### **Example 4:** Changing memory resource escalation strategy

```bash
process {
    withName: 'ASSEMBLY' {
        cpus    = 8
        memory  = { check_max( escalate_exp( 64.GB, task, 2 ), 'memory' ) }
        queue = 'long'
        time    = 48.h
    }
}
```
Ensure to add the folowing functions copied from the `nextflow-commons.config` to the `resource_tweak config`: 
1) `def check_max { ... }`
2) `def escalate_exp { ... }`

> 

Reviewer 2: Can we just add these functions in here rather than then having to go digging for it?

> 1) `def check_max { ... }`
> 2) `def escalate_exp { ... }`
> 

Reviewer 2: ```suggestion
1) `def check_max { ... }`
2) `def escalate_exp { ... }`
3)  `def escalate_linear { ... }`

```

Giving the user the option to use a linear retry strategy as well.
